File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        while (true) {
            String line = br.readLine();
            if (line.equals(".")) break;
            
            if (isBalanced(line)) {
                System.out.println("yes");
            } else {
                System.out.println("no");
            }
        }
        br.close();
    }
    
    public static boolean isBalanced(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (c == '(' || c == '[') {
                stack.push(c);
            } else if (c == ')') {
                if (stack.isEmpty() || stack.pop() != '(') {
                    return false; 
                }
            } else if (c == ']') {
                if (stack.isEmpty() || stack.pop() != '[') {
                    return false; 
                }
            }
        }
        
        return stack.isEmpty(); // 스택이 비어 있어야 균형 잡힘
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
from sys import stdin
input = stdin.readline


def solution():
    def check(line):
        stack = []
        dic = {
            "(": -1,
            ")": 1,
            "[": -2,
            "]": 2
        }
        for l in line:
            if l in ["(", "["]:
                stack.append(l)
            elif l in [")", "]"]:
                if not stack:
                    print("no")
                    return
                prev = stack.pop()
                if dic[prev] + dic[l] != 0:
                    print("no")
                    return
        print("yes") if not stack else print("no")
    while True:
        l = input().rstrip()
        if l[0] == ".":
            return
        check(l)


if __name__ == "__main__":
    solution()
